### DISCUSSION TITLE
concentration zero cut off in solver compared to experiment

### DISCUSSION THREAD

**Discussion by dion-w** [pybamm-team/PyBaMM#1982] [Q&A] [answered]
Hello everyone,
i saw the solver operates on events to terminate the simulation. I don't question the event of the cutoff-voltage but what about the zero concentration cut-offs?
Speaking from an experiment point of view the experiment will always reach the voltage cut-off since there are some limiting processes which increase the potential losses in the cell. In the simulation though it can also terminate due to the electrolyte concentration or the concentration in the solid being smaller than zero. Since i know concentrations smaller zero make no sense, a diffusion coefficient or conductivity (in the electrolyte) could go to zero (or almost zero such as 1e-8 mol/m³ for numerical reasons) to increase the potential loss in this phase. This way the potential loss of this process would rise and the voltage cut off would be met, just like in the experiment. The profit of this would be to achieve better comparability to experiments!

From experience i already know that capping the concentration to very low values works, e.g. it worked for our institute-own matlab solver of a P2D model, although by doing so you would introduce a small amount of charge into system. This may be small enough though (e.g. when using very low concentrations such as 1e-8 mol/m³) to not alter the system too much when dealing with a single discharge but it might have to be evaluated for cycling.

So my idea would be to drop the events for concentration cut-offs in the solver or make them optional and add the very low finite amount of concentration as capping so there is always a very minimal amount of concentration present in the system to not have singularities. Maybe some extrapolation for transport coefficients so they go against zero can also be added just like it is the case for the OCV-curves?

What do you think about it or do you have already better ideas? At least this what we came up with last year when we dealt with the same problem. We only looked at single discharges in thick electrodes where the electrolyte transportation is the limiting process.

**Comment by brosaplanella** [comment]
Hi @dion-w! This is a very relevant point. At the moment, the standard DFN should work fine if we drop the event (at least for the cases I tested it worked fine). However, there are still some errors arising for the degradation models (I got some errors with plating). In any case, it would be good to see what other people think and if we should get rid of the electrolyte depletion event.

**Comment by valentinsulzer** [comment, marked as answer]
Yeah, we can probably remove it for the DFN (or just make it smaller than the correction for numerics).

There's two ways of having the correction: applied to the variable that goes into the equation, or applied afterwards (so the variable in the equation could go negative, but the variable that goes into the exchange-current is always positive). Which did you do in your matlab script, @dion-w ?

For SPMe, we need to keep the event cut-off, since there is no "self-correction" to keep c_e above 0 like there is in the DFN (at least, until we implement the "SPMe with depletion")

**Reply by dion-w** [reply]
We're solving our matlab P2D model with ode15s and integrating only the time domain whereas we are creating our right hand side with our own spatial discretization. The function with creates the rhs sets every negative concentration value to 1e-8 mol/m³ and then proceeds to calculate to rhs of the current timesteps. So the variables in the solver could get negative but they get set to a small positive value every step. This was merely a workaround at that point. There is an option in the ode15s to have some variables not being allowed to become negative but we de didn't try it. It may be more stable though.

I also think c_e only being positive is not only important for the exchange current density but also for the equation of transport in the electrolyte. There you have the dln(c_e)/dx term for which also c_e must not become negative

**Reply by dion-w** [reply]
@tinosulzer Since you mentioned it i thought about the "SPMe with depletion" the last few days: I think the SPMe can not give accurate results in this setup due to the modeling aproximations of representing the whole electrode as a single particle. After all this means that the electrode is homogeneously lithiated. But If you're in a scenaria where you have severe depletion in your electrolyte then you typically deal with thick electrodes or high C-rates (or even both). The electrolyte depletion creates high gradients of the electrolyte concentration which would lead to a strong variation of lithiation across the thickness of the battery. This violates the assumption of representing the electrode as homogeneously lithiated and therefore as a single particle. 

Of course theoretically speaking, an "SPMe with depletion" could make sense when you have thin porous electrodes but a very dense and/or tortous separator. But I can't think of a scenario where you would use such a system. The separator would be the source of the highest ohmic voltage loss and therefore the cell's capacity couldn't be used fully.

I mean, i don't know if there is more to the idea of the "SPMe with depletion" but this is what i understood for now.

**Reply by valentinsulzer** [reply]
Yeah that's a good point. Could we have an SPMe with two representative particles, one for non-depleted and one for depleted? I guess the one in depleted would never change concentration and the front would just move? I'm just waffling here but it seems like there should be a way to do it rigorously.
